### PR TITLE
add exception handling

### PR DESCRIPTION
**Pull request by Meitoseshifu**

### CODE REVIEW

File: src/main/java/com/bobocode/petros/container/ApplicationAnnotationContainer.java [hunk 1]
```diff
@@ -35,20 +33,26 @@ public ApplicationAnnotationContainer(String packageName) {
...
     public <T> T getDependency(Class<T> clazz) {
-        if (isNonUniqueDependency(clazz)) {
-            LOG.debug("The dependency with such {} type already exists in container", clazz.getName());
-            throw new NoUniqueDependecyException(clazz.getName());
-        } else {
-            return getDependencyFromMap(clazz);
+        T dependency = null;
+        try {
+            dependency = getDependencyFromMap(clazz);
+        } catch (NoUniqueDependencyException e) {
```
Comment (StanislavZabramnyi) [inline src/main/java/com/bobocode/petros/container/ApplicationAnnotationContainer.java:52]: method `getDependencyFromMap` can also throw NosuchElementException . I guess we should pass it to this level and handle it with a meaningful message, not just a default one -  "no value present'. What do you think?

**Review by StanislavZabramnyi — APPROVED** [review]

File: src/main/java/com/bobocode/petros/container/ApplicationAnnotationContainer.java (hunk 1 above)
Comment (Meitoseshifu) [reply src/main/java/com/bobocode/petros/container/ApplicationAnnotationContainer.java:52]: Geeeeez, I've missed that, thanks)

**Review by Markysss — APPROVED** [review]

**Review by Markysss — COMMENTED** [review]
Good job!

File: src/main/java/com/bobocode/petros/container/ApplicationAnnotationContainer.java [hunk 2]
```diff
@@ -66,9 +70,13 @@ private DependencyDefinition keyByDependencyDefinitionName(String name) {
 
     @SuppressWarnings("unchecked")
     private <T> T getDependencyFromMap(Class<T> clazz) {
-        return (T) dependencyMap.values().stream()
-                .filter(obj -> obj.getClass().equals(clazz))
-                .findFirst()
-                .orElseThrow();
+        if(!isNonUniqueDependency(clazz)) {
```
Comment (yura-jt) [inline src/main/java/com/bobocode/petros/container/ApplicationAnnotationContainer.java:73]: `if(!`
not formatted code - missing space

File: src/main/java/com/bobocode/petros/container/ApplicationAnnotationContainer.java [hunk 3]
```diff
@@ -66,9 +70,13 @@ private DependencyDefinition keyByDependencyDefinitionName(String name) {
...
+            return (T) dependencyMap.values().stream()
+                    .filter(obj -> obj.getClass().equals(clazz))
+                    .findFirst()
+                    .orElseThrow();
+        } else {
```
Comment (yura-jt) [inline src/main/java/com/bobocode/petros/container/ApplicationAnnotationContainer.java:78]: If and else is +1 level of abstraction, usually if possible trying to get rid of it.
For example in this case:

```
private <T> T getDependencyFromMap(Class<T> clazz) {
  if (!isNonUniqueDependency(clazz)) {
     throw new NoUniqueDependencyException(clazz.getName());
  }
  
  return (T) dependencyMap.values().stream()
                    .filter(obj -> obj.getClass().equals(clazz))
                    .findFirst()
                    .orElseThrow();
}
```
logically it should be the same, but in term of reading and maintaining it's more easier.

**Review by yura-jt — APPROVED** [review]

**Review by Shtramak — APPROVED** [review]

File: src/main/java/com/bobocode/petros/container/ApplicationAnnotationContainer.java (hunk 3 above)
Comment (Meitoseshifu) [reply src/main/java/com/bobocode/petros/container/ApplicationAnnotationContainer.java:78]: Yeah, nice catch)

**Pull request merged**